x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/utils/InferenceProcessorInfoExtractorTests.java

```diff
@@ -0,0 +1,212 @@
+                        put(
+                            "on_failure",
+                            Arrays.asList(inferenceProcessorForModel("second_level"), forEachProcessorWithInference("third_level"))
+                        );
+                    }
+                }))))
+        ) {
+            configurations.put(
+                "pipeline_with_model_nested",
+                new PipelineConfiguration("pipeline_with_model_nested", BytesReference.bytes(xContentBuilder), XContentType.JSON)
+            );
+        }
+
+        IngestMetadata ingestMetadata = new IngestMetadata(configurations);
+
+        ClusterState cs = ClusterState.builder(new ClusterName("_name"))
+            .metadata(Metadata.builder().putCustom(IngestMetadata.TYPE, ingestMetadata))
+            .nodes(
+                DiscoveryNodes.builder()
+                    .add(new DiscoveryNode("min_node", new TransportAddress(InetAddress.getLoopbackAddress(), 9300), Version.CURRENT))
+                    .add(new DiscoveryNode("current_node", new TransportAddress(InetAddress.getLoopbackAddress(), 9302), Version.CURRENT))
+                    .add(new DiscoveryNode("_node_id", new TransportAddress(InetAddress.getLoopbackAddress(), 9304), Version.CURRENT))
+                    .localNodeId("_node_id")
+                    .masterNodeId("_node_id")
+            )
+            .build();
+
+        assertThat(InferenceProcessorInfoExtractor.countInferenceProcessors(cs), equalTo(3));
+    }
+
+    private static PipelineConfiguration newConfigurationWithOutInferenceProcessor(int i) throws IOException {
+        try (
+            XContentBuilder xContentBuilder = XContentFactory.jsonBuilder()
+                .map(
+                    Collections.singletonMap(
+                        "processors",
+                        Collections.singletonList(Collections.singletonMap("not_inference", Collections.emptyMap()))
+                    )
+                )
+        ) {
+            return new PipelineConfiguration("pipeline_without_model_" + i, BytesReference.bytes(xContentBuilder), XContentType.JSON);
+        }
+    }
+
+    private static ClusterState buildClusterState(Metadata metadata) {
+        return ClusterState.builder(new ClusterName("_name")).metadata(metadata).build();
+    }
+
+    private static ClusterState buildClusterStateWithModelReferences(int numPipelineReferences, String... modelId) throws IOException {
+        Map<String, PipelineConfiguration> configurations = Maps.newMapWithExpectedSize(modelId.length);
+        for (String id : modelId) {
+            for (int i = 0; i < numPipelineReferences; i++) {
+                configurations.put(
+                    "pipeline_with_model_" + id + i,
+                    randomBoolean() ? newConfigurationWithInferenceProcessor(id) : newConfigurationWithForeachProcessorProcessor(id)
+                );
+            }
+
+        }
+        for (int i = 0; i < randomInt(5); i++) {
```

Comment:
It's confusing to regenerate the upper bound on each loop iteration, and means there's very little chance of ever getting to 4.  It would be better to generate the upper bound once, before the loop.

Reply:
Ah, definitely.